apps/cowswap-frontend/src/widgetEventEmitter.ts

```diff
@@ -1,5 +1,210 @@
+      orderId: payload.orderUid,
+      chainId: payload.chainId,
+      sellToken: safeGetString(payload.inputToken, 'address'),
+      buyToken: safeGetString(payload.outputToken, 'address'),
+      sellAmount: payload.inputAmount?.toString() || '',
+      buyAmount: payload.outputAmount?.toString() || '',
+    }
+  }
+
+  if (isOrderPayload(payload)) {
+    return {
+      walletAddress: safeGetString(payload.order, 'owner'),
+      orderId: safeGetString(payload.order, 'uid'),
+      chainId: payload.chainId,
+      sellToken: safeGetString(payload.order, 'sellToken'),
+      buyToken: safeGetString(payload.order, 'buyToken'),
+      sellAmount: safeGetString(payload.order, 'sellAmount'),
+      buyAmount: safeGetString(payload.order, 'buyAmount'),
+      ...(payload.transactionHash ? { transactionHash: payload.transactionHash } : {}),
+    }
+  }
+
+  // Fallback for unexpected payload types
+  return createDefaultTrackingPayload()
+}
+
+// Helper to set up an event handler with common logic
+const setupEventHandler = <T extends CowWidgetEvents>(
+  event: T,
+  getAdditionalProps: (payload: any) => Record<string, any>,
+  isPostedOrder = false,
+) => {
+  WIDGET_EVENT_EMITTER.on({
+    event,
+    handler: (payload: unknown) => {
+      const eventName = EVENT_MAPPING[event]
+      if (eventName) {
+        const analytics = getCowAnalytics()
+        if (analytics) {
+          const commonProps = getCommonOrderProperties(payload, isPostedOrder)
+          const additionalProps = getAdditionalProps(payload)
+
+          analytics.sendEvent(eventName, {
+            ...commonProps,
+            ...additionalProps,
+          })
+        } else {
+          console.warn('Analytics instance not available for event:', eventName)
+        }
+      }
+    },
+  })
+}
+
+const setupEventHandlers = () => {
+  // Handle order submission
+  setupEventHandler(
+    CowWidgetEvents.ON_POSTED_ORDER,
+    (payload) => ({
+      sellTokenSymbol: safeGetString(payload.inputToken, 'symbol'),
+      buyTokenSymbol: safeGetString(payload.outputToken, 'symbol'),
+      orderType: payload.orderType,
+      kind: payload.kind,
+      receiver: payload.receiver || '',
+      orderCreationHash: payload.orderCreationHash || '',
+    }),
+    true, // isPostedOrder
+  )
+
+  // Handle order fulfillment
+  setupEventHandler(CowWidgetEvents.ON_FULFILLED_ORDER, (payload) => ({
+    executedSellAmount: safeGetString(payload.order, 'executedSellAmount'),
+    executedBuyAmount: safeGetString(payload.order, 'executedBuyAmount'),
+    executedFeeAmount: safeGetString(payload.order, 'executedFeeAmount'),
+  }))
+
+  // Handle order cancellation
+  setupEventHandler(CowWidgetEvents.ON_CANCELLED_ORDER, (payload) => ({
+    reason: 'cancelled',
+    transactionHash: payload.transactionHash || '',
+  }))
+
+  // Handle order expiration
+  setupEventHandler(CowWidgetEvents.ON_EXPIRED_ORDER, () => ({
+    reason: 'expired',
+  }))
+}
+
+setupEventHandlers()
```

Comment:
The analytics logic is supposed to be in `libs/analytics`

Reply:
You can export `setupEventHandlers` from the lib and call it here as `setupEventHandlers(WIDGET_EVENT_EMITTER)`